pkg/apis/backupdriver/v1/snapshot.go

```diff
@@ -0,0 +1,148 @@
+package v1
+
+import meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type SnapshotSpec struct {
+	// ResourceHandle refers to a Kubernetes resource, currently a PVC but this may be extended in the future
+	ResourceHandle string
+
+	// The backup repository to snapshot into.  The namespace the Snapshot/PVC lives in must have access to the repository
+	BackupRepository string
+}
+
+// SnapshotPhase represents the lifecycle phase of a Snapshot.
+// New - No work yet, next phase is InProgress
+// InProgress - snapshot being taken
+// Snapshotted - local snapshot complete, next phase is Protecting or SnapshotFailed
+// SnapshotFailed - end state, snapshot was not able to be taken
+// Uploading - snapshot is being moved to durable storage
+// Uploaded - end state, snapshot has been protected
+// UploadFailed - end state, unable to move to durable storage
+// Canceling - when the SanpshotCancel flag is set, if the Snapshot has not already moved into a terminal state, the
+//             status will move to Canceling.  The snapshot ID will be removed from the status status if has been filled in
+//             and the snapshot ID will not longer be valid for a Clone operation
+// Canceled - the operation was canceled, the snapshot ID is not valid
+type SnapshotPhase string
+
+const (
+	SnapshotPhaseNew            SnapshotPhase = "New"
+	SnapshotPhaseInProgress     SnapshotPhase = "InProgress"
+	SnapshotPhaseSnapshotted    SnapshotPhase = "Snapshotted"
+	SnapshotPhaseSnapshotFailed SnapshotPhase = "SnapshotFailed"
+	SnapshotPhaseUploading      SnapshotPhase = "Uploading"
+	SnapshotPhaseUploaded       SnapshotPhase = "Uploaded"
+	SnapshotPhaseUploadFailed   SnapshotPhase = "UploadFailed"
+	SnapshotPhaseCanceling      SnapshotPhase = "Canceling"
+	SnapshotPhaseCanceled       SnapshotPhase = "Canceled"
+)
```

Comment:
Do we need to track Uploading/Upload/UploadFailed as part of snapshot? That will be tracked by the Upload CR.

Reply:
This is the only status that the Guest Cluster sees.  If we split the data movement out separately in the Backup Driver APi then it would make sense to move the status to the data movement API but currently it's all in one.

Reply:
Who is responsible for updating the status between upload CR and snapshot CR? Alternately, can the Upload CR live in the same namespace as the snapshot CR? That way, the pvBackup is responsible for getting it into the guest. If we add new fields to upload CR, we do not have to replicate it in the snapshot Spec to inform the guest.

Reply:
The Backup Driver is currently responsible for updating from the Upload CR->Snapshot CR.  I think where it gets a little funky is with the guest/supervisor crossing.

We create a Snapshot CR in the Guest Cluster NS.  The Backup Driver is responsible for seeing it, and translating that to a Supervisor Cluster Snapshot CR.  If we then write an Upload CR in Guest Cluster, something will need to translate it to the Supervisor Cluster, probably Backup Driver.  So we're going to wind up echoing the status somehow.

Reply:
Ok. Guess the details will become clear as we implement. Thanks for explaining.